### ISSUE TITLE
Feature request: Flag to specify a preset file

### ISSUE THREAD

**Issue by ad-si** [mixn/carbon-now-cli#31]
This CLI is very ... uhm ... special. None of the obvious flags do exist 🙈

`cat my-preset.json`

```json
{
    "t": "monokai",
    "bg": "#ADB7C1",
    "wt": "sharp",
    "wc": true,
    "fm": "Hack",
    "fs": "18px",
    "ln": false,
    "ds": true,
    "dsyoff": "20px",
    "dsblur": "68px",
    "wa": true,
    "lh": "133%",
    "pv": "16",
    "ph": "24",
    "si": false,
    "wm": false,
    "es": "2x",
    "type": "png"
}
```

```shell
carbon-now --preset-path=./my-preset.json test.js
```

**Comment by mixn** [comment]
Hi @ad-si, thanks for chiming in.

This comment is very… uhm… special. None of the clarifications on what you actually mean do exist. 🙈😁

It is anything but special to have one global config file in `~/`, if that is what you are trying to say.

<img width="1165" alt="screen shot 2018-09-18 at 14 56 47" src="https://user-images.githubusercontent.com/672237/45689036-31facd80-bb53-11e8-9b65-0995afe4083e.png">

I’m planning to stick with one global `~/.carbon-now.json`, at least for now.

**Comment by mixn** [comment]
@ad-si I’ve thought about this some more and having a hard time seeing the benefits of adding a flag like this. 

`carbon-now-cli` functions on a per-preset basis. Meaning, every preset is kind of like a config in its own right.

It is similar to something like [Yeoman’s `.yo-rc.json`](http://yeoman.io/authoring/storage.html), where each generator configuration is namespaced. Hence, only one config file is needed.

It is different to something like [terminalizer’s per directory structure](https://github.com/faressoft/terminalizer#config), which functions on a per-config basis. And even then there’s a [need for global configs](https://github.com/faressoft/terminalizer/issues/4).

I’d appreciate if you could elaborate on why this seems useful.

**Comment by ad-si** [comment]
I'm generating some screenshots for the docs of a project of mine.
If somebody else want's to hack on that project as well, they can't simply run `make`, because they first have to copy the preset into their own `~/.carbon-now.json` file.
Hence, there should be a way to specify a local preset file which gets bundled with the project.

If you don't like skipping the preset name it could also be realized as:

`my-preset.json`:
```json
{
  "ProjectPreset": {
    "t": "monokai",
    "bg": "#ADB7C1",
    "wt": "sharp",
    "wc": true,
    "fm": "Hack",
    "fs": "18px",
    "ln": false,
    "ds": true,
    "dsyoff": "20px",
    "dsblur": "68px",
    "wa": true,
    "lh": "133%",
    "pv": "16",
    "ph": "24",
    "si": false,
    "wm": false,
    "es": "2x",
    "type": "png"
  }
}
```
```shell
carbon-now --preset-file=./my-preset.json --preset=ProjectPreset test.js
```

**Comment by ad-si** [comment]
Haha, well then ... 😄

> What do you think should the shorthand for --preset-file be?

Like I said it in the other issue: I wouldn't add a shorthand
Especially not for this flag, as it will probably mostly be used in scripts.

**Comment by mixn** [comment]
@ad-si Wouldn’t a `--config` be more desirable/fitting? Since

```
{
  "ProjectPreset": {
    "t": "monokai",
    "bg": "#ADB7C1",
    "wt": "sharp",
    "wc": true,
    "fm": "Hack",
    "fs": "18px",
    "ln": false,
    "ds": true,
    "dsyoff": "20px",
    "dsblur": "68px",
    "wa": true,
    "lh": "133%",
    "pv": "16",
    "ph": "24",
    "si": false,
    "wm": false,
    "es": "2x",
    "type": "png"
  }
}
```

essentially **is** a config. It’d behave the same way as `~/.carbon-now.json` without auto-saving `latest-preset` to it to not modify the git history when run via a script.

Thoughts?

**Comment by mixn** [comment]
> But I guess it should be without the preset name than

Wouldn’t that be even more confusing, since it would a) differ from the usual config and b) limit it to only one preset instead of `x` via `carbon-now --config=./local-config.json --preset=preset-57`? One config with the same structure allowing limitless presets and mixtures.

> Actually I don't even know myself right now without looking it up

I don’t either, took it straight from Carbon’s URL params to not having to map the map again. 😄

**Comment by ad-si** [comment]
Yeah, not sure either. Probably `carbon-now --config=./local-config.json --preset=preset-57` is fine. I guess consistency between `local-config.json` and `carbon-now.json` is a good thing.
But what happens when you don't specify a preset? Should it throw an error?

> params to not having to map the map again

Great UX has it's price ;-)

**Comment by mixn** [comment]
> But what happens when you don't specify a preset? Should it throw an error?

I’d rather keep the `latest-preset` behaviour then. Which will result to it being written every time you run it and have `--config` set. 🤔 Not sure.

> Great UX has its price 😉

This was in the very early stages, way before presets. I might re-map, but not a high priority atm.

**Comment by mixn** [comment]
@ad-si I’ve thought about this some more and think this is the best approach:

```
carbon-now foo.js --config=./local-config.json --preset=something
```

It behaves the same way as using the global config, for the only difference that it:

1. Will not write `latest-preset` to `local-config.json`, file stays untouched.
2. It also won’t create `local-config.json` if it hasn’t existed before.

Not specifying `--preset` would result in the default settings being used (as it is right now).

This allows for `x` presets inside the local config, doesn’t modify Git history for third-party projects using `carbon-now-cli` via scripts and keeps the API in sync.

**Comment by ad-si** [comment]
Sounds good!
But maybe it should print a warning / error if a config file is specified, but not used? Or is there a reason why anyone should do it without it being an error?

**Comment by mixn** [comment]
@ad-si I just published `v1.3.0-0` where this is included. 🙂 Docs are [here](https://github.com/mixn/carbon-now-cli#local-configs) + warning for non-existent presets included. 👍 

<img width="747" alt="screen shot 2018-09-26 at 16 12 46" src="https://user-images.githubusercontent.com/672237/46085803-29d00d00-c1a7-11e8-9a48-b1f3b538ce1d.png">
